ui/components/forms/flavors/compound-form/_index.scss

```diff
@@ -21,7 +21,10 @@
   }
 
   .slds-form-element {
-    width: 100%; // Default - Stretch 100% unless told otherwise using sizing helpers
+    // Default - Stretch 100% unless told otherwise using sizing helpers
+    &:not([class*="slds-size--"]) {
+      width: 100%;
+    }
```

Comment:
This approach would work fine if the responsive classes would be set as BEM elements, `slds-size--x-of-y`, `slds-size__medium--x-of-y`, `slds-size__xx-large--x-of-y`...